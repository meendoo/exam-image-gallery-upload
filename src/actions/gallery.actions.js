import React from 'react'
import { GALLERY } from '../constants'
import { firestore } from '../services/firebase';
import { showLoading, hideLoading } from 'react-redux-loading-bar'
import { toast } from 'react-toastify';
import { ImageActions } from '../actions'

export const fetchImages = () => {
    return async dispatch => {
        dispatch(showLoading());
        dispatch(fetchRequest());
        try {
            await firestore.collection('images')
                .onSnapshot(snapshot => {
                    let fetchedData = {
                        images: []
                    }
                    snapshot.forEach(doc => {
                        fetchedData.images.push({
                            id: doc.id,
                            ...doc.data(),
                        });
                    });
                    dispatch(fetchSuccess(fetchedData.images))
                });
        } catch (error) {
            toast.error(error);
            dispatch(hideLoading());
        }
        dispatch(hideLoading());
    }
};

export const fetchRequest = () => {
    return { type: GALLERY.FETCH_REQUEST };
}

export const fetchSuccess = (images) => {
    return { type: GALLERY.FETCH_SUCCESS, images };
}

export const orderByNewest = () => {
    return { type: GALLERY.ORDERBY_NEWEST }
};

export const orderByOldest = () => {
    return { type: GALLERY.ORDERBY_OLDEST }
};

export const ToastCloseButton = ({ closeToast, confirmAction }) => {
    const toastStyle = {width: '48px', height: '48px', border: 'none', borderRadius: '0', background: "#b71d1d", color: "white", fontWeight: 'bold', cursor: 'pointer'};
    return (
        <button style={toastStyle} onClick={confirmAction} autoFocus>YES</button>
    )
}

export const clearGallery = (refs) =>
    async dispatch => {
        toast.error("Delete all images?", {
            position: 'top-right',
            closeButton: <ToastCloseButton confirmAction={confirmDelete}/>
        });
        function confirmDelete() {
            refs.forEach(ref => {                
                dispatch(ImageActions.deleteImage(ref))
            });
        } 
        return { type: GALLERY.CLEAR_GALLERY }
}