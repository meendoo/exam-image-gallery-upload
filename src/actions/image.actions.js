import { IMAGE } from '../constants'
import { firestore, storage } from '../services/firebase';

export const viewImage = (imageRef, currentViewIndex) => {
    return { type: IMAGE.VIEW_IMAGE, payload: {imageRef, currentViewIndex} }
}

export const deleteImage = (imageRef) => async dispatch => {
    try {
        firestore.collection('images').doc(imageRef.id).delete();
        storage.ref().child(`images/${imageRef.name}`).delete();
    } catch (error) {
        console.log(error);
    }
    dispatch({ type: IMAGE.DELETE });
}