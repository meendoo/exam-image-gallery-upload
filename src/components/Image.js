import React, { Component } from "react";
import styles from "./image.module.scss";
import CaptionInput from "./CaptionInput";

export default class Image extends Component {
  constructor(props) {
    super(props);
    this.state = {
      isCaptionEditable: false
    };
  }

  toggleInputCaption = (e, state) => {
    e.stopPropagation();
    this.setState({
      isCaptionEditable: state
    });
  };

  render() {
    const { imageId, updateImageCaption, className, url, caption } = this.props;
    const { isCaptionEditable } = this.state;
    return (
      <figure className={`${styles.figure} ${className}`}>
        <div
          style={{ backgroundImage: `url('${url}')` }}
          className={`${styles.imgDiv} ${className}`}
        />
        <figcaption className={styles.figcaption} onClick={e => this.toggleInputCaption(e, true)}>
          {!isCaptionEditable ? (
            <p>{String(caption).substr(0, 250)}</p>
          ) : (
            <CaptionInput
              caption={caption}
              imageId={imageId}
              hideInput={e => this.toggleInputCaption(e, false)}
              updateImageCaption={updateImageCaption}
            />
          )}
        </figcaption>
      </figure>
    );
  }
}
