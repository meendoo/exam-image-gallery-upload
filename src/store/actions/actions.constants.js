export const GALLERY = {
    IS_EMPTY: '@GALLERY_IS_EMPTY',
    LIST_IMAGES: '@GALLERY_LIST_IMAGES',
    ORDERBY_NEWEST: '@GALLERY_ORDERBY_NEWEST',
    ORDERBY_OLDEST: '@GALLERY_ORDERBY_OLDEST',
    ZOOM_IN: '@GALLERY_ZOOM_IN',
    ZOOM_OUT: '@GALLERY_ZOOM_OUT'
}

export const MODAL = {
    OPEN: '@MODAL_OPEN',
    CLOSE: '@MODAL_CLOSE',
    SHOWPREVIEW: '@MODAL_SHOWPREVIEW',
    HIDEPREVIEW: '@MODAL_HIDEPREVIEW',
    FILES_DROPPED: '@MODAL_FILES_DROPPED',
    CLEAR_FILES: '@MODAL_CLEAR_FILES'
}

export const IMAGE = {
    ROTATE: '@IMAGE_ROTATE',
    DELETE: '@IMAGE_DELETE'
}