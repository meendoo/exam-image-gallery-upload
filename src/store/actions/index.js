export * from './gallery.actions';
export * from './modal.actions';
export * from './image.actions';
