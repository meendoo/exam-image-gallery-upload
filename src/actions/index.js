import * as GalleryActions from './gallery.actions';
import * as ModalActions from './modal.actions';
import * as ImageActions from './image.actions';

export { GalleryActions };
export { ModalActions };
export { ImageActions };
