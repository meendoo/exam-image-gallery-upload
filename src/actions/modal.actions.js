import { MODAL } from '../constants'
import { storage, firestore } from '../services/firebase';
import { toast } from "react-toastify";
import { ImageActions } from '../actions'

const successMessage = files => `Awesome! ${files.length} image${files.length > 1 ? 's' : ''} ${files.length > 1 ? 'were' : 'was'} uploaded successfully.`

export const openUploadZoneModal = () => {
    return { type: MODAL.OPEN_UPLOAD_ZONE };
}

export const closeModal = () => {
    return { type: MODAL.CLOSE_MODAL };
}

export const openImageViewerModal = (imageRef, currentViewIndex) => dispatch => { 
    dispatch({ type: MODAL.OPEN_IMAGE_VIEWER });
    dispatch(ImageActions.viewImage(imageRef, currentViewIndex));
}

export const sendFiles = (files) => dispatch => {
    let storagePromises = [];
    dispatch(closeModal());
    dispatch({ type: MODAL.UPLOAD_REQUEST });
    for (let i = 0; i < files.length; i++) {
        let toastId = null;
        storage.ref().child(`dummybckp/${files[i].name}`).put(files[i]); // for monitoring purposes
        let fbStorage = storage.ref().child(`images/${files[i].name}`).put(files[i]);
        storagePromises.push(fbStorage);
        fbStorage.on('state_changed',
            snapshot => {
                let progress = (snapshot.bytesTransferred / snapshot.totalBytes);
                if(toastId === null){
                    toastId = toast('Upload in Progress', {progress: progress, progressClassName: 'toastProgressBar'});
                } else {
                    toast.update(toastId, {progress: progress});
                }
            },
            err => {
                toast.error(err, {toastId: toastId});
            },
            async () => {
                toast.done(toastId);
                fbStorage.snapshot.ref.getDownloadURL()
                    .then(imgURL => {
                        firestore.collection('images').add({
                            "url": imgURL,
                            "name": fbStorage.snapshot.metadata.name,
                            "timestamp": new Date(),
                            "orientation": 0
                        })
                    }).catch(err => toast.error(err));
            }
        );
    };

    Promise.all(storagePromises)
        .then(()=> {
            dispatch({ type: MODAL.UPLOAD_SUCCESS });
            toast.done();
            toast.success(successMessage(files), {autoClose: true})
        })
        .catch(err => toast.error(err));
}