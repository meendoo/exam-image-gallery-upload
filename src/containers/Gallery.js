import React, { Component } from "react";
import { connect } from "react-redux";
import Image from "../components/Image";
import styles from "./gallery.module.scss";
import { TimelineMax } from "gsap";
import { ModalActions, GalleryActions } from "../actions";
import Spinner from "../components/Spinner";

export class Gallery extends Component {
  constructor(props) {
    super(props);
    this.imageAnimation = new TimelineMax();
  }

  animateImages = () => {
    return this.imageAnimation.staggerFromTo(
      `.${styles.image}`,
      0.2,
      { autoAlpha: 0 },
      { autoAlpha: 1 },
      0.05
    );
  };

  componentDidUpdate = prevProps => {
    if (
      prevProps.images !== this.props.images ||
      prevProps.order !== this.props.order
    ) {
      this.animateImages().play();
    }
  };

  componentDidMount = () => {
    this.props.fetchImages();
  };

  render() {
    this.imageAnimation.kill().clear();
    const { images } = this.props;
    const imgs =
      images &&
      images.length > 0 &&
      images.map((img, imgIndex) => {
        return (
          <Image
            key={img.imageId}
            url={img.url}
            className={styles.image}
            onClick={() => this.props.openImageViewerModal(images, imgIndex)}
          />
        );
      });

    return (
      <>
        {this.props.isUploading && <Spinner />}
        {images && images.length > 0 ? (
          <div className={styles.gallery}>{imgs}</div>
        ) : (
          <div className={styles.placeholderText}>
            You haven't uploaded images yet :(
          </div>
        )}
      </>
    );
  }
}

const mapStateToProps = ({ gallery, modal }) => ({
  images: gallery.images,
  isUploading: modal.isUploading
});

const mapDispatchToProps = dispatch => ({
  fetchImages: () => dispatch(GalleryActions.fetchImages()),
  openImageViewerModal: (imageRef, currentViewIndex) =>
    dispatch(ModalActions.openImageViewerModal(imageRef, currentViewIndex))
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Gallery);
