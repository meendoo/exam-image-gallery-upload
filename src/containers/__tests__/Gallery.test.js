import React from "react";
import Gallery from "../Gallery";
import Image from "../../components/Image";
import { GalleryActions } from "../../actions/index";
import GALLERY from "../../constants/gallery.constants";
import { shallow } from "enzyme";
import sinon from "sinon";
import configureMockStore from "redux-mock-store";
import thunk from "redux-thunk";
import renderer from "react-test-renderer";

const initialState = {
  gallery: {
    images: [
      {
        imageId: 0,
        name: "A.jpeg",
        timestamp: {
          seconds: 1548900000
        },
        url: "https://picsum.photos/200/300"
      },
      {
        imageId: 1,
        name: "B.jpeg",
        timestamp: {
          seconds: 1548800000
        },
        url: "https://picsum.photos/300/400"
      },
      {
        imageId: 2,
        name: "C.jpeg",
        timestamp: {
          seconds: 1548700000
        },
        url: "https://picsum.photos/400/500"
      }
    ]
  },
  modal: {
    isUploading: false
  }
};

describe("Connected <Gallery/> with mockStore", () => {
  let store;
  const mockStore = configureMockStore([thunk]);

  beforeEach(() => {
    store = mockStore(initialState);
  });

  const createImages = store => {
    let storeImages = store.getState().gallery.images;
    return storeImages.map((img, imgIndex) => (
      <Image
        key={img.imageId}
        url={img.url}
        onClick={() => this.props.openImageViewerModal(images, imgIndex)}
      />
    ));
  };

  it("should show 3 images when provided", () => {
    const images = createImages(store);
    const galleryWithProps = shallow(<Gallery store={store}>{images}</Gallery>);
    expect(galleryWithProps.children().length).toEqual(3);
    expect(galleryWithProps).toMatchSnapshot();
  });

  it("it should render as it is", () => {
    const galleryInitial = renderer.create(<Gallery store={store} />).toJSON();
    expect(galleryInitial).toMatchSnapshot();
  });

  // it("should play animation if images and order props are changed", () => {
  //   const spy = sinon.spy(Gallery.prototype.componentDidUpdate);
  //   const wrapper = shallow(<Gallery store={store} />);

  //   store.dispatch({ type: GALLERY.FETCH_SUCCESS, images: [] });
  //   console.log(store.getActions());

  //   // expect(spy).toBeCalled();
  //   // expect(spy).to.have.property("callCount", 0);
  //   // wrapper.setProps({ foo: 'foo' });
  //   // expect(spy).to.have.property('callCount', 1);
  // });
});
