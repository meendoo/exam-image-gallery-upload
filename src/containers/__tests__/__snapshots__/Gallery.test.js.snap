// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Connected <Gallery/> with mockStore it should render as it is 1`] = `
<div
  className="gallery"
>
  <figure
    className="figure image"
    onClick={[Function]}
  >
    <div
      className="imgDiv image"
      style={
        Object {
          "backgroundImage": "url('https://picsum.photos/200/300')",
        }
      }
    />
  </figure>
  <figure
    className="figure image"
    onClick={[Function]}
  >
    <div
      className="imgDiv image"
      style={
        Object {
          "backgroundImage": "url('https://picsum.photos/300/400')",
        }
      }
    />
  </figure>
  <figure
    className="figure image"
    onClick={[Function]}
  >
    <div
      className="imgDiv image"
      style={
        Object {
          "backgroundImage": "url('https://picsum.photos/400/500')",
        }
      }
    />
  </figure>
</div>
`;

exports[`Connected <Gallery/> with mockStore should show 3 images when provided 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Connect(Gallery)
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
  >
    <Image
      onClick={[Function]}
      url="https://picsum.photos/200/300"
    />
    <Image
      onClick={[Function]}
      url="https://picsum.photos/300/400"
    />
    <Image
      onClick={[Function]}
      url="https://picsum.photos/400/500"
    />
  </Connect(Gallery)>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": Array [
        <Image
          onClick={[Function]}
          url="https://picsum.photos/200/300"
        />,
        <Image
          onClick={[Function]}
          url="https://picsum.photos/300/400"
        />,
        <Image
          onClick={[Function]}
          url="https://picsum.photos/400/500"
        />,
      ],
      "fetchImages": [Function],
      "images": Array [
        Object {
          "imageId": 0,
          "name": "A.jpeg",
          "timestamp": Object {
            "seconds": 1548900000,
          },
          "url": "https://picsum.photos/200/300",
        },
        Object {
          "imageId": 1,
          "name": "B.jpeg",
          "timestamp": Object {
            "seconds": 1548800000,
          },
          "url": "https://picsum.photos/300/400",
        },
        Object {
          "imageId": 2,
          "name": "C.jpeg",
          "timestamp": Object {
            "seconds": 1548700000,
          },
          "url": "https://picsum.photos/400/500",
        },
      ],
      "isUploading": false,
      "openImageViewerModal": [Function],
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "storeSubscription": Subscription {
        "listeners": Object {
          "clear": [Function],
          "get": [Function],
          "notify": [Function],
          "subscribe": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": [Function],
      },
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": "0",
        "nodeType": "function",
        "props": Object {
          "onClick": [Function],
          "url": "https://picsum.photos/200/300",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "1",
        "nodeType": "function",
        "props": Object {
          "onClick": [Function],
          "url": "https://picsum.photos/300/400",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "2",
        "nodeType": "function",
        "props": Object {
          "onClick": [Function],
          "url": "https://picsum.photos/400/500",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <Image
            onClick={[Function]}
            url="https://picsum.photos/200/300"
          />,
          <Image
            onClick={[Function]}
            url="https://picsum.photos/300/400"
          />,
          <Image
            onClick={[Function]}
            url="https://picsum.photos/400/500"
          />,
        ],
        "fetchImages": [Function],
        "images": Array [
          Object {
            "imageId": 0,
            "name": "A.jpeg",
            "timestamp": Object {
              "seconds": 1548900000,
            },
            "url": "https://picsum.photos/200/300",
          },
          Object {
            "imageId": 1,
            "name": "B.jpeg",
            "timestamp": Object {
              "seconds": 1548800000,
            },
            "url": "https://picsum.photos/300/400",
          },
          Object {
            "imageId": 2,
            "name": "C.jpeg",
            "timestamp": Object {
              "seconds": 1548700000,
            },
            "url": "https://picsum.photos/400/500",
          },
        ],
        "isUploading": false,
        "openImageViewerModal": [Function],
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "storeSubscription": Subscription {
          "listeners": Object {
            "clear": [Function],
            "get": [Function],
            "notify": [Function],
            "subscribe": [Function],
          },
          "onStateChange": [Function],
          "parentSub": undefined,
          "store": Object {
            "clearActions": [Function],
            "dispatch": [Function],
            "getActions": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
          },
          "unsubscribe": [Function],
        },
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": "0",
          "nodeType": "function",
          "props": Object {
            "onClick": [Function],
            "url": "https://picsum.photos/200/300",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "1",
          "nodeType": "function",
          "props": Object {
            "onClick": [Function],
            "url": "https://picsum.photos/300/400",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "2",
          "nodeType": "function",
          "props": Object {
            "onClick": [Function],
            "url": "https://picsum.photos/400/500",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
